pallets/randomness/src/lib.rs

```diff
@@ -0,0 +1,231 @@
+                    // TODO: add logic to check parent relay block (ideally, we make it valid for `curr_relay_block - 2`)
+                    let latest_valid_block = frame_system::Pallet::<T>::block_number()
+                        .saturating_sub(sp_runtime::traits::One::one());
+
+                    // Save it to be readily available for use
+                    LatestBabeRandomness::<T>::put((randomness, latest_valid_block));
+
+                    // Emit an event detailing that a new randomness is available
+                    Self::deposit_event(Event::NewRandomnessAvailable {
+                        randomness_seed: randomness,
+                        from_epoch: relay_epoch_index,
+                        valid_until_block: latest_valid_block,
+                    });
+                } else {
+                    log::warn!(
+                        "Failed to fill BABE epoch randomness \
+							REQUIRE HOTFIX TO FILL EPOCH RANDOMNESS FOR EPOCH {:?}",
+                        relay_epoch_index
+                    );
+                }
+            }
+
+            // Update storage with the latest epoch for which randomness was processed for
+            <RelayEpoch<T>>::put(relay_epoch_index);
+
+            // Update storage to reflect that this inherent was included in the block (so the block is valid)
+            <InherentIncluded<T>>::put(());
+
+            // Inherents do not pay for execution
+            Ok(Pays::No.into())
+        }
+    }
+
+    /// Implement the required trait to provide an inherent to the runtime
+    #[pallet::inherent]
+    impl<T: Config> ProvideInherent for Pallet<T> {
+        type Call = Call<T>;
+        type Error = InherentError;
+        const INHERENT_IDENTIFIER: InherentIdentifier = INHERENT_IDENTIFIER;
+
+        // This function returns if the inherent should be added to the current block or not
+        fn is_inherent_required(_: &InherentData) -> Result<Option<Self::Error>, Self::Error> {
+            // Return Ok(Some(_)) unconditionally because this inherent is required in every block
+            // If it is not found, throw a InherentRequired error.
+            Ok(Some(InherentError::Other(
+                sp_runtime::RuntimeString::Borrowed("Inherent required to set babe randomness"),
+            )))
+        }
+
+        // The empty-payload inherent extrinsic.
+        fn create_inherent(_data: &InherentData) -> Option<Self::Call> {
+            Some(Call::set_babe_randomness {})
+        }
+
+        fn is_inherent(call: &Self::Call) -> bool {
+            matches!(call, Call::set_babe_randomness { .. })
+        }
+    }
+
+    #[pallet::hooks]
+    impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
+        /// This hook checks, on block finalization, that the required inherent was included and clears
+        /// storage to make it necesary to include it in future blocks as well
+        fn on_finalize(_now: BlockNumberFor<T>) {
```

Comment:
IIRC, the weight used by `on_finalize` had to be defined in `on_initialize`, so that the block producer can know if it will be able to produce a block. I know this literally only does one rad, and modification of the chain's state, but since there is no `on_initialize` here, how does it work?

Reply:
Yes, forgot the `on_initialize` hook that returns the `on_finalize` weight. Fixed (although we still have to benchmark `on_finalize`)